### PR TITLE
Clickid localstorage

### PR DESCRIPTION
**Pull request by prak24oct**

### CODE REVIEW

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/ReferrerDetails.kt [hunk 1]
```diff
@@ -20,7 +20,12 @@ data class RefererDetails(
     init {
         val decodedUrl = URLDecoder.decode(url, "UTF-8").toString()
         params = Util.getQueryParams(decodedUrl);
-        clickId = params["tr_clickid"] ?: ""
+        if(params.containsKey("clickId")) {
+            clickId = params["tr_clickid"] ?: ""
+        }
+        else{
+            clickId = url
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/ReferrerDetails.kt:27]: why is clickId having URL?

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+class LocalInstallReferrer {
+    var clickId = ""
+    fun getLocalRefDetails(delimeter:String): RefererDetails {
+        try{
+        File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).absolutePath).walk()
+            .forEach {
+                val pattern = Regex("^[a-bA-Z0-9-_]*_[a-f\\d]{24}\$")
+                if (pattern.containsMatchIn(it.nameWithoutExtension)) {
+                    val value = it.nameWithoutExtension.split(delimeter)
+                    Log.d("openChooser", "apk: :" + value.get(value.size - 1))
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt:20]: remove log

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                    clickId = value.get(value.size - 1)
+                }
+            }
+        return RefererDetails(clickId, "", "")
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt:24]: indentation

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import com.android.installreferrer.api.InstallReferrerStateListener
+import kotlinx.coroutines.delay
+import java.io.File
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+class LocalInstallReferrer {
+    var clickId = ""
+    fun getLocalRefDetails(delimeter:String): RefererDetails {
+        try{
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt:14]:
```suggestion
        try {
```

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import android.util.Log
+import com.android.installreferrer.api.InstallReferrerStateListener
+import kotlinx.coroutines.delay
+import java.io.File
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+class LocalInstallReferrer {
+    var clickId = ""
+    fun getLocalRefDetails(delimeter:String): RefererDetails {
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt:13]: delimiter should be set in constructor

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt (hunk 2 above)
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt:20]: remove log

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+                val pattern = Regex("^[a-bA-Z0-9-_]*_[a-f\\d]{24}\$")
+                if (pattern.containsMatchIn(it.nameWithoutExtension)) {
+                    val value = it.nameWithoutExtension.split(delimeter)
+                    Log.d("openChooser", "apk: :" + value.get(value.size - 1))
+                    clickId = value.get(value.size - 1)
+                }
+            }
+        return RefererDetails(clickId, "", "")
+        } catch (e: Exception) {
+            return return RefererDetails("", "", "")
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt:26]:
```suggestion
            return RefererDetails.default()
```

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+            val rd = getLocalRefDetails(delimeter)
+            it.resume(rd)
+        }
+    }
+    
+    suspend fun getRefDetails(delimeter:String): RefererDetails {
+        for (i in 1..5) {
+            try {
+                return getInfo(delimeter)
+            } catch (ex: InstallReferrerException) {
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt:41]: exception is not thrown anywhere so it can be removed

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/ReferrerDetails.kt [hunk 8]
```diff
@@ -20,7 +20,12 @@ data class RefererDetails(
     init {
         val decodedUrl = URLDecoder.decode(url, "UTF-8").toString()
         params = Util.getQueryParams(decodedUrl);
-        clickId = params["tr_clickid"] ?: ""
+        if(params.containsKey("clickId")) {
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/ReferrerDetails.kt:23]: undo the changes

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt [hunk 9]
```diff
@@ -0,0 +1,47 @@
+package com.trackier.sdk
+
+import android.os.Environment
+import android.util.Log
+import com.android.installreferrer.api.InstallReferrerStateListener
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt:5]: import is not used anywhere

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+import kotlinx.coroutines.delay
+import java.io.File
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+class LocalInstallReferrer {
+    var clickId = ""
+    fun getLocalRefDetails(delimeter:String): RefererDetails {
+        try{
+        File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).absolutePath).walk()
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt:15]: we should check for permissions before calling the function @prak24oct

**Review by Hemant-Mann — CHANGES_REQUESTED** [review]

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/ReferrerDetails.kt (hunk 1 above)
Comment (prak24oct) [reply TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/ReferrerDetails.kt:27]: In earlier case we are getting install referee url from which  we are extracting clickId in param, but here we are getting direct clicked

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/TrackierSDKInstance.kt [hunk 11]
```diff
@@ -131,9 +133,15 @@ class TrackierSDKInstance {
             return
         }
         if (!isReferrerStored()) {
-            val installRef = InstallReferrer(this.config.context)
-            val refDetails = installRef.getRefDetails()
-            this.setReferrerDetails(refDetails)
+            if(isLocalRefEnabled) {
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/TrackierSDKInstance.kt:136]:
```suggestion
            if (isLocalRefEnabled) {
```

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+class LocalInstallReferrer(val context: Context, val delimeter: String) {
+    var clickId = ""
+    @RequiresApi(Build.VERSION_CODES.M)
+    fun getLocalRefDetails(): RefererDetails {
+        try {
+            if (context.checkSelfPermission(READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
+                File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).absolutePath).walk()
+                    .forEach {
+                        val pattern = Regex("^[a-bA-Z0-9-_]*_[a-f\\d]{24}\$")
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/LocalInstallReferrer.kt:21]:
```suggestion
                        val pattern = Regex("[a-fA-F\\d]{24}\$")
```

**Review by Hemant-Mann — APPROVED** [review]
